coderd/database/sqlxqueries/bindvars.go

```diff
@@ -0,0 +1,95 @@
+package sqlxqueries
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+	"strings"
+
+	"golang.org/x/xerrors"
+
+	"github.com/google/uuid"
+	"github.com/jmoiron/sqlx/reflectx"
+	"github.com/lib/pq"
+
+	"github.com/coder/coder/coderd/util/slice"
+)
+
+var nameRegex = regexp.MustCompile(`@([a-zA-Z0-9_]+)`)
+
+// dbmapper grabs struct 'db' tags.
+var dbmapper = reflectx.NewMapper("db")
+
+// bindNamed is an implementation that improves on the SQLx implementation. This
+// adjusts the query to use "$#" syntax for arguments instead of "@argument". The
+// returned args are the values of the struct fields that match the names in the
+// correct order and indexing.
+//
+// 1. SQLx does not reuse arguments, so "@arg, @arg" will result in two arguments
+// "$1, $2" instead of "$1, $1".
+// 2. SQLx does not handle uuid arrays.
+// 3. SQLx only supports ":name" style arguments and breaks "::" type casting.
+func bindNamed(query string, arg interface{}) (newQuery string, args []interface{}, err error) {
```

Comment:
Is there no other way around this? I really dislike the regexp rewriting and reflection on every single query.

Is sqlx not smarter than this?

Reply:
Actually SQLx is a bit dummber. I took this code implementation from SQLx and changed some things. The usage for SQLx is `sqlx.Named()`. But it doesn't handle things that I put in the function comments, so I adjusted it.

https://github.com/jmoiron/sqlx/blob/master/named.go#LL421-L436C2

Instead of a regex, they wrote a more proper parser: https://github.com/jmoiron/sqlx/blob/master/named.go#L331

But the parser doesn't work. It parses comments and they use `:arg` as arguments, so when we typecast with `::`, it's seen as escaping a colon literal. So our sql queries would have to do `::::`.

The SQLx parser is honestly really dumb with actual SQL.